streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -43,7 +42,7 @@
 import static org.apache.kafka.streams.StreamsConfig.InternalConfig.EMIT_INTERVAL_MS_KSTREAMS_OUTER_JOIN_SPURIOUS_RESULTS_FIX;
 import static org.apache.kafka.streams.processor.internals.metrics.TaskMetrics.droppedRecordsSensor;
 
-class KStreamKStreamJoin<K, V1, V2, VOut> implements ProcessorSupplier<K, V1, K, VOut> {
+abstract class KStreamKStreamJoin<K, VL, VR, VOut, VThis, VOther> implements ProcessorSupplier<K, VThis, K, VOut> {
```

Comment:
Also a better name for `VL` and `VR` would be `VLeft` and `VRight`, respectively. `VR` is sometimes used for result value type.

Reply:
We needed a way to detect the type of the side being called dynamically. So that the super class can handle them based on the caller.

When this class is called by the _left side_ of the join `VThis` would be _left_ left type (`VLeft`) and `VOther` would be _right_ type (`VRight`).

```java
class KStreamKStreamLeftJoinSide<K, VLeft, VRight, VOut> extends KStreamKStreamJoin<K, VLeft, VRight, VOut, VLeft, VRight> {
																											--^--  --^---
																											VThis VOther
```

Whereas for the _right side_ of the join `VThis` would be _right_ left type (`VRight`) and `VOther` would be _left_ type (`VLeft`).

```java
class KStreamKStreamRightJoinSide<K, VLeft, VRight, VOut> extends KStreamKStreamJoin<K, VLeft, VRight, VOut, VRight, VLeft>
																											--^--  --^---
																											VThis VOther
```

Reply:
This is more tangible in the abstract methods:
```java
protected abstract VThis getThisValue(final LeftOrRightValue<? extends VLeft, ? extends VRight> leftOrRightValue);

protected abstract VOther getOtherValue(final LeftOrRightValue<? extends VLeft, ? extends VRight> leftOrRightValue);
```
Depending on which join side implements these methods, `VThis` and `VOther` can change. For instance, the getThisValue on the _left side_ of the join would be:
```java
 @Override
public VLeft getThisValue(final LeftOrRightValue<? extends VLeft, ? extends VRight> leftOrRightValue) {
      return leftOrRightValue.getLeftValue();
}
```
Respectively, on the _right side_ we have
```java
  @Override
  public VRight getThisValue(final LeftOrRightValue<? extends VLeft, ? extends VRight> leftOrRightValue) {
      return leftOrRightValue.getRightValue();
  }
  ````

Reply:
Just to throw my own explanation in here:
VLeft and VRight are "absolute" in that they are the left and right types of the overall join. Both sides of the join have equivalent VLeft and VRight types, because they share a common outerJoinStore instance.

VThis and VOther are "relative" in that they are the type of records entering "this" side of the join, and the "other" side of the join, and this is necessarily swapped for the other side of the join.